gtsfm/densify/mvsnets/methods/PatchmatchNet/models/patchmatch.py

```diff
@@ -0,0 +1,891 @@
+        """forward
+        Args:
+            ref_feature: reference feature map
+            grid: position of sampling points in adaptive spatial cost aggregation
+        """
+        batch, feature_channel, height, width = ref_feature.size()
+
+        x = F.grid_sample(ref_feature, grid, mode="bilinear", padding_mode="border")
+
+        # [B,G,C//G,H,W]
+        ref_feature = ref_feature.view(batch, self.G, feature_channel // self.G, height, width)
+
+        x = x.view(batch, self.G, feature_channel // self.G, self.neighbors, height, width)
+        # [B,G,Neighbor,H,W]
+        x = (x * ref_feature.unsqueeze(3)).mean(2)
+        del ref_feature
+        # [B,Neighbor,H,W]
+        x = self.similarity(self.conv1(self.conv0(x))).squeeze(1)
+
+        return self.output(x)
+
+
+def depth_weight(
+    depth_sample: torch.Tensor,
+    depth_min: int,
+    depth_max: int,
+    grid: torch.Tensor,
+    patchmatch_interval_scale: float,
+    evaluate_neighbors: int,
+) -> torch.Tensor:
+    """calculate depth weight
+    adaptive spatial cost aggregation
+    weight based on depth difference of sampling points and center pixel
+    Args:
+        depth_sample: sample depth map
+        depth_min: minimum virtual depth
+        depth_max: maximum virtual depth
+        grid: position of sampling points in adaptive spatial cost aggregation
+        patchmatch_interval_scale: patchmatch interval scale,
+        evaluate_neighbors: number of evaluate neighbors
+    Returns:
+        depth weight
+    """
+    neighbors = evaluate_neighbors
+    batch, num_depth, height, width = depth_sample.size()
+    # normalization
+    x = 1.0 / depth_sample
+    del depth_sample
+    inverse_depth_min = 1.0 / depth_min
+    inverse_depth_max = 1.0 / depth_max
+    x = (x - inverse_depth_max.view(batch, 1, 1, 1)) / (
+        inverse_depth_min.view(batch, 1, 1, 1) - inverse_depth_max.view(batch, 1, 1, 1)
+    )
+
+    x1 = F.grid_sample(x, grid, mode="bilinear", padding_mode="border")
+    del grid
+    x1 = x1.view(batch, num_depth, neighbors, height, width)
+
+    # [B,Ndepth,N_neighbors,H,W]
+    x1 = torch.abs(x1 - x.unsqueeze(2)) / patchmatch_interval_scale
+    del x
+    x1 = torch.clamp(x1, min=0, max=4)
+    # sigmoid output approximate to 1 when x=4
+    x1 = (-x1 + 2) * 2
+    output = nn.Sigmoid()
+    x1 = output(x1)
+
+    return x1.detach()
+
+
+class PixelwiseNet(nn.Module):
+    """Pixelwise Net
+    estimate pixel-wise view weight
+    """
+
+    def __init__(self, G: int):
+        """initialize
+        G: number of groups
+        """
+        super(PixelwiseNet, self).__init__()
+        self.conv0 = ConvBnReLU3D(G, 16, 1, 1, 0)
+        self.conv1 = ConvBnReLU3D(16, 8, 1, 1, 0)
+        self.conv2 = nn.Conv3d(8, 1, kernel_size=1, stride=1, padding=0)
+        self.output = nn.Sigmoid()
+
+    def forward(self, x1: torch.Tensor) -> torch.Tensor:
+        """forward

```

Comment:
This docstring could also be clearer -- feed what type of input through what type of network? What is the meaning of the output?